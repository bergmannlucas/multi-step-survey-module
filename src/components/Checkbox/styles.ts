import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin: 1rem 0;
`

export const OptionWrapper = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-gap: 15px;
`

export const Input = styled.input`
  width: 2rem;
  height: 2rem;
  cursor: pointer;
`

export const Label = styled.label`
  ${({ theme }) => css`
    margin-bottom: 1rem;
    font-size: ${theme.font.sizes.xlarge};
    font-weight: ${theme.font.bold};
    color: ${theme.colors.purple};
  `}
`

export const Option = styled.span`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.xlarge};
    color: ${theme.colors.purple};
    line-height: normal;
  `}
`
