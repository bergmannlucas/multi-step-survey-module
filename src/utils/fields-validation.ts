import Joi from 'joi'

export const schema = Joi.array().items(
  Joi.object({
    name: Joi.string().empty(''),
    email: Joi.string()
      .email({ tlds: { allow: false } })
      .empty('')
  }),
  Joi.object({
    gender: Joi.string().required(),
    age: Joi.number().required()
  }),
  Joi.object({
    book: Joi.string().required(),
    colors: Joi.object({
      black: Joi.boolean(),
      white: Joi.boolean(),
      red: Joi.boolean(),
      green: Joi.boolean(),
      blue: Joi.boolean(),
      yellow: Joi.boolean(),
      purple: Joi.boolean(),
      orange: Joi.boolean(),
      pink: Joi.boolean()
    })
  })
)
