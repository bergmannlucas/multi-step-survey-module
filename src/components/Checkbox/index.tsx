import { InputHTMLAttributes } from 'react'

import * as S from './styles'

type CheckboxValue = {
  [key: string]: boolean
}

type CheckboxProps = {
  labelFor?: string
  options?: string[] | number[]
  label?: string
  selectedValues?: CheckboxValue[]
} & InputHTMLAttributes<HTMLInputElement>

function Checkbox({
  labelFor = '',
  options,
  label,
  selectedValues,
  ...props
}: CheckboxProps) {
  return (
    <S.Wrapper>
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}
      <S.OptionWrapper>
        {options?.map((option) => (
          <div key={option}>
            <S.Input
              id={labelFor}
              type="checkbox"
              value={option}
              checked={!!selectedValues![option as number]}
              {...props}
            />
            <S.Option>{option}</S.Option>
          </div>
        ))}
      </S.OptionWrapper>
    </S.Wrapper>
  )
}

export default Checkbox
