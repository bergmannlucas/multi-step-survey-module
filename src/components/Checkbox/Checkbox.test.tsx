import userEvent from '@testing-library/user-event'
import { render, screen } from 'utils/test-utils'
import Checkbox from '.'

describe('<Checkbox />', () => {
  it('should render with label', () => {
    const { container } = render(
      <Checkbox
        options={['black']}
        selectedValues={[{ black: true }]}
        label="checkbox label"
        labelFor="check"
        onChange={jest.fn()}
      />
    )

    expect(screen.getByRole('checkbox')).toBeInTheDocument()
    expect(screen.getByLabelText(/checkbox label/i)).toBeInTheDocument()
    expect(screen.getByText(/checkbox label/i)).toHaveAttribute('for', 'check')

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render without label', () => {
    render(<Checkbox />)

    expect(screen.queryByLabelText('Checkbox')).not.toBeInTheDocument()
  })

  it('should be accessible with tab', () => {
    render(
      <Checkbox
        options={['black']}
        selectedValues={[{ black: true }]}
        label="checkbox label"
        labelFor="check"
        onChange={jest.fn()}
      />
    )

    expect(document.body).toHaveFocus()

    userEvent.tab()

    expect(screen.getByLabelText(/checkbox/i)).toHaveFocus()
  })
})
