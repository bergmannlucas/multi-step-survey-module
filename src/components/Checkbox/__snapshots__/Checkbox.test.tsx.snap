// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render with label 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 1rem 0;
}

.c2 {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-gap: 15px;
}

.c3 {
  width: 2rem;
  height: 2rem;
  cursor: pointer;
}

.c1 {
  margin-bottom: 1rem;
  font-size: 2.0rem;
  font-weight: 700;
  color: #9249FF;
}

.c4 {
  font-size: 2.0rem;
  color: #9249FF;
  line-height: normal;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="check"
  >
    checkbox label
  </label>
  <div
    class="c2"
  >
    <div>
      <input
        class="c3"
        id="check"
        type="checkbox"
        value="black"
      />
      <span
        class="c4"
      >
        black
      </span>
    </div>
  </div>
</div>
`;
